### PR TITLE
[W-10463851] Make Donation History component compatible with DE.

### PR DESCRIPTION
**Pull request by mcasco**
# Work Item
[W-10463851](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000h1a3yYAA/view)

# Critical Changes

Make Donation History components work in the context of a Community taking into account the currently connected user.
( these changes depend on the permissions discussed in the ticket [W-10281560](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000WJIN2YAP/view) )

# Changes

- Check the existence of the contact id when initializing the component.
- Add method to return the contact id related to the current user.
- enable component targets in the meta to be able to use them in a DE.

### CODE REVIEW

File: force-app/main/default/classes/DonationHistoryController.cls [hunk 1]
```diff
@@ -69,6 +69,12 @@ public with sharing class DonationHistoryController {
             this.paymentMethod = paymentMethod;
         }
     }
+
+    public class DonationHistoryAppView {
+        @AuraEnabled public Boolean isPaymentsEnabled;
+        @AuraEnabled public List<DonationHistoryView> donations;
+        @AuraEnabled public String paymentMethodLabel;
+    }
```
Comment (bdvorachek) [inline force-app/main/default/classes/DonationHistoryController.cls:77]: This can be combined with the DonationHistoryAppView defined on lines 42-47.

**Review by bdvorachek — COMMENTED** [review]
There's a bit of extra code (Gift.cls etc) from a @psadie-sfdo PR merge that happened upstream showing up in this PR. I'm going to fix that by pulling all the changes from the feature/238 branch down into feature/238__refactor. This branch already has those changes somehow, which is why they're showing up in the PR review. Looks like the auto-merge functionality may have failed to run here.

**Comment by JoseRossiSF** [comment]
This PR also includes the required Permission Set described in [W-10281560](https://gus.lightning.force.com/a07EE00000WJIN2YAP)

File: force-app/main/default/permissionsets/NPSP_Donation_History_Community_User_managed.permissionset-meta.xml [hunk 2]
```diff
@@ -0,0 +1,89 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (bdvorachek) [inline force-app/main/default/permissionsets/NPSP_Donation_History_Community_User_managed.permissionset-meta.xml:1]: Check with @jencrabtree on the naming of this permission set. I don't think we need to include the word "managed" in this perm set.

File: force-app/main/default/lwc/donationHistoryTable/donationHistoryTable.js [hunk 3]
```diff
@@ -35,6 +37,12 @@ export default class DonationHistoryTable extends LightningElement {
     @track
     columns = [];
 
+    // eslint-disable-next-line @lwc/lwc/no-async-await
+    async connectedCallback() {
```
Comment (bdvorachek) [inline force-app/main/default/lwc/donationHistoryTable/donationHistoryTable.js:41]: The connectedCallback here does not need to be `async`.

File: force-app/main/default/lwc/donationHistoryTable/donationHistoryTable.js [hunk 4]
```diff
@@ -63,16 +71,13 @@ export default class DonationHistoryTable extends LightningElement {
...
-    connectedCallback() {
-        this.retrieveDonationHistory();
+        this.tableElement ? this.tableElement.enableInfiniteLoading = true : null;
     }
 
     // eslint-disable-next-line @lwc/lwc/no-async-await
-    async retrieveDonationHistory() {
-        getDonationHistory({contactId: this.contactId, offset: 0, filter : this.filter})
-        .then(data => {
+    async retrieveDonationHistory(){
```
Comment (bdvorachek) [inline force-app/main/default/lwc/donationHistoryTable/donationHistoryTable.js:78]: This function does not need to be `async` if we're using promises inside of the function body instead of `await`.

File: force-app/main/default/lwc/givingSummaryForDetailPage/givingSummaryForDetailPage.js [hunk 5]
```diff
@@ -1,7 +1,18 @@
 import { api, LightningElement } from 'lwc';
+import getContactIdByUserId from "@salesforce/apex/DonationHistoryController.getContactIdByUserId";
 
 export default class GivingSummaryDetailPage extends LightningElement {
     
     @api recordId;
 
+    // eslint-disable-next-line @lwc/lwc/no-async-await
+    async connectedCallback() {
```
Comment (bdvorachek) [inline force-app/main/default/lwc/givingSummaryForDetailPage/givingSummaryForDetailPage.js:9]: This function does not need to be `async`.

File: force-app/main/default/permissionsets/NPSP_Donation_History_Community_User_managed.permissionset-meta.xml (hunk 2 above)
Comment (JoseRossiSF) [reply force-app/main/default/permissionsets/NPSP_Donation_History_Community_User_managed.permissionset-meta.xml:1]: It has been decided the Permission Set will be removed from this PR

File: force-app/main/default/classes/DonationHistoryController_TEST.cls [hunk 6]
```diff
@@ -92,7 +93,17 @@ public class DonationHistoryController_TEST {
         List<Opportunity> recordsOpp = new List<Opportunity> {opportunityHardCreditOfContact0, opportunityAnotherHardCreditOfContact0, opportunityForContact1, anotherOpportunityForContact1};
         // recordsOpp.add(opportunityHardCredit);
         // recordsOpp.add(opportunitySoftCredit);
+
         insert recordsOpp;
+        
+                if (!Schema.sObjectType.Profile.isAccessible()) {
```
Comment (dcaballeroSf) [inline force-app/main/default/classes/DonationHistoryController_TEST.cls:99]: Indentation is wrong

File: force-app/main/default/classes/USER_UserService.cls [hunk 7]
```diff
@@ -256,6 +256,21 @@ public with sharing class USER_UserService {
...
+    * @description method to return contact related to current user.
+    */
+    public static Id getContactIdByUserId() {
+        Id userId = UserInfo.getUserId();
+        List<User> contactId = [SELECT ContactId
+                                FROM User
+                                WHERE Id =: userId
+                                WITH SECURITY_ENFORCED
+                                LIMIT 1];
+        return contactId.get(0).ContactId;
```
Comment (dcaballeroSf) [inline force-app/main/default/classes/USER_UserService.cls:271]: You could do:
User theUser = [SELECT.... WHERE Id =: userId LIMIT 1]

Also contactId is not a proper name as you are getting an unser from the query

File: force-app/main/default/lwc/donationHistoryFilter/donationHistoryFilter.js [hunk 8]
```diff
@@ -18,7 +19,12 @@ export default class DonationHistoryFilter extends LightningElement {
     value = LIFETIME_VALUE;
     
     connectedCallback() {
-        getYearsWithDonation({contactId : this.contactId})
+        this.getYearsByContact();
+    }
+
+    getYearsByContact(){
```
Comment (dcaballeroSf) [inline force-app/main/default/lwc/donationHistoryFilter/donationHistoryFilter.js:25]: Why did you change it? I dont see the point and also the name getYearsByContact is more confusing than getYearsWithDonations. If you wanted to rename it, it would be better getYearsWithDonationsByContact

File: force-app/main/default/lwc/givingSummaryForDetailPage/givingSummaryForDetailPage.js [hunk 9]
```diff
@@ -1,7 +1,17 @@
 import { api, LightningElement } from 'lwc';
+import getContactIdByUserId from "@salesforce/apex/DonationHistoryController.getContactIdByUserId";
 
 export default class GivingSummaryDetailPage extends LightningElement {
     
     @api recordId;
 
+    connectedCallback() {
+        if(!this.recordId){
```
Comment (dcaballeroSf) [inline force-app/main/default/lwc/givingSummaryForDetailPage/givingSummaryForDetailPage.js:9]: spacing here

**Review by dcaballeroSf — CHANGES_REQUESTED** [review]
A few styling comments

File: force-app/main/default/classes/DonationHistoryController.cls [hunk 10]
```diff
@@ -77,7 +78,8 @@ public with sharing class DonationHistoryController {
     */
     @AuraEnabled(cacheable=true)
     public static DonationHistoryAppView getDonationHistory(Id contactId, Integer offset, String filter) {
-        DonationHistoryService service = new DonationHistoryService(contactId, offset, filter);
+        contactId = (contactId == null ? getContactIdByUserId() : contactId );
+		DonationHistoryService service = new DonationHistoryService(contactId, offset, filter);
```
Comment (lparrott) [inline force-app/main/default/classes/DonationHistoryController.cls:82]: Fix indentation here, makes it seem like one long statement.

File: force-app/main/selector/DonationHistorySelector.cls [hunk 11]
```diff
@@ -125,9 +125,12 @@ public inherited sharing class DonationHistorySelector {
         return totalRecords;
     }
 
+
+
     /** 
     * @author Salesforce.org
     * @date 2021
+
```
Comment (lparrott) [inline force-app/main/selector/DonationHistorySelector.cls:133]: Extra line here, also we normally only include the "author" and "date" at the top of Classes.

File: force-app/main/default/classes/DonationHistoryController_TEST.cls (hunk 6 above)
Comment (bdvorachek) [reply force-app/main/default/classes/DonationHistoryController_TEST.cls:99]: I don't think we need to check for access to Profiles as part of test data setup. The running user for a unit test should be a sysadmin usually, who would have access to profiles. If we need to test other levels of access like a standard user, we can mock those permissions in the test.

File: force-app/main/default/lwc/givingSummaryForDetailPage/givingSummaryForDetailPage.js [hunk 12]
```diff
@@ -1,7 +1,17 @@
 import { api, LightningElement } from 'lwc';
+import getContactIdByUserId from "@salesforce/apex/DonationHistoryController.getContactIdByUserId";
 
 export default class GivingSummaryDetailPage extends LightningElement {
```
Comment (bdvorachek) [inline force-app/main/default/lwc/givingSummaryForDetailPage/givingSummaryForDetailPage.js:4]: The class name here does not line up with the name of the actual LWC. Can we change the class name to match the file/folder name of GivingSummaryForDetailPage?

File: force-app/main/default/classes/DonationHistoryController.cls (hunk 10 above)
Comment (JoseRossiSF) [reply force-app/main/default/classes/DonationHistoryController.cls:82]: Hi @lparrott. I do not understand the suggestion. To what exact statement are you referring to?

Comment (bdvorachek) [reply force-app/main/default/classes/DonationHistoryController.cls:82]: Line 82 is indented with two tab characters instead of 8 spaces. We generally prefer to indent using 4 spaces per level of indentation.

**Review by lparrott — APPROVED** [review]

**Review by bdvorachek — APPROVED** [review]

**Review by JoseRossiSF — APPROVED** [review]
